File: lib/common/utils/phi_detector.dart
/// PHI (Protected Health Information) Detection Utility
/// Helps prevent accidental inclusion of patient identifiable information
/// in reflections and CPD entries

class PhiDetector {
  // Common PHI patterns to detect
  static final List<RegExp> _patterns = [
    // NHS Number (10 digits)
    RegExp(r'\b\d{3}\s?\d{3}\s?\d{4}\b', caseSensitive: false),
    
    // Names that look like full names (Mr/Mrs/Ms/Dr + First Last or First Middle Last)
    RegExp(r'\b(Mr|Mrs|Ms|Dr|Miss|Prof|Professor)\.?\s+[A-Z][a-z]+\s+[A-Z][a-z]+\b', caseSensitive: false),
    
    // Patient identifiers
    RegExp(r'\bpatient\s+(?:id|number|identifier)\s*:?\s*\w+', caseSensitive: false),
    RegExp(r'\bMRN\s*:?\s*\w+', caseSensitive: false), // Medical Record Number
    
    // Dates of birth
    RegExp(r'\bDOB\s*:?\s*\d{1,2}[/-]\d{1,2}[/-]\d{2,4}\b', caseSensitive: false),
    RegExp(r'\bborn\s+(?:on\s+)?\d{1,2}[/-]\d{1,2}[/-]\d{2,4}\b', caseSensitive: false),
    
    // Addresses (partial - postcode patterns)
    RegExp(r'\b[A-Z]{1,2}\d{1,2}\s?\d[A-Z]{2}\b'), // UK postcode
    RegExp(r'\b\d+\s+[A-Z][a-z]+\s+(Street|Road|Avenue|Lane|Drive|Close|Way)\b', caseSensitive: false),
    
    // Phone numbers (UK format)
    RegExp(r'\b0\d{10}\b'),
    RegExp(r'\b\+44\s?\d{10}\b'),
    
    // Email addresses
    RegExp(r'\b[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}\b'),
    
    // Hospital numbers
    RegExp(r'\b[Hh]ospital\s+[Nn]umber\s*:?\s*\w+', caseSensitive: false),
    
    // Ward/bed numbers (common patterns)
    RegExp(r'\bward\s+[A-Z0-9]+\b', caseSensitive: false),
    RegExp(r'\bbed\s+\d+', caseSensitive: false),
  ];
  
  // High-risk keywords
  static final List<String> _keywords = [
    'patient name',
    'patient\'s name',
    'nhs number',
    'date of birth',
    'hospital number',
    'medical record',
    'home address',
    'phone number',
    'email address',
  ];
  
  /// Check text for potential PHI
  /// Returns list of detected PHI types
  static List<String> detectPhi(String text) {
    final detected = <String>{};
    final lowerText = text.toLowerCase();
    
    // Check for pattern matches
    for (final pattern in _patterns) {
      if (pattern.hasMatch(text)) {
        detected.add(_getPatternDescription(pattern));
      }
    }
    
    // Check for keywords
    for (final keyword in _keywords) {
      if (lowerText.contains(keyword)) {
        detected.add('Keyword: "$keyword"');
      }
    }
    
    return detected.toList();
  }
  
  /// Check if text contains potential PHI
  static bool containsPhi(String text) {
    return detectPhi(text).isNotEmpty;
  }
  
  /// Get warning level based on number of PHI indicators
  static PhiWarningLevel getWarningLevel(String text) {
    final detected = detectPhi(text);
    
    if (detected.isEmpty) return PhiWarningLevel.none;
    if (detected.length == 1) return PhiWarningLevel.low;
    if (detected.length <= 3) return PhiWarningLevel.medium;
    return PhiWarningLevel.high;
  }
  
  static String _getPatternDescription(RegExp pattern) {
    final patternStr = pattern.pattern;
    
    if (patternStr.contains('NHS') || patternStr.contains(r'\d{3}\s?\d{3}\s?\d{4}')) {
      return 'Possible NHS Number';
    }
    if (patternStr.contains('Mr|Mrs|Ms|Dr')) {
      return 'Possible patient name';
    }
    if (patternStr.contains('DOB') || patternStr.contains('born')) {
      return 'Possible date of birth';
    }
    if (patternStr.contains('@')) {
      return 'Email address';
    }
    if (patternStr.contains('postcode') || patternStr.contains('[A-Z]{1,2}\\d{1,2}')) {
      return 'Possible address/postcode';
    }
    if (patternStr.contains('0\\d{10}') || patternStr.contains('\\+44')) {
      return 'Phone number';
    }
    if (patternStr.contains('ward') || patternStr.contains('bed')) {
      return 'Ward/bed identifier';
    }
    
    return 'Potential PHI detected';
  }
  
  /// Get user-friendly warning message
  static String getWarningMessage(List<String> detected) {
    if (detected.isEmpty) return '';
    
    final buf = StringBuffer();
    buf.writeln('⚠️ Potential Patient Information Detected\n');
    buf.writeln('The following may contain identifiable patient information:\n');
    
    for (final item in detected) {
      buf.writeln('• $item');
    }
    
    buf.writeln('\n📋 NHS IG Guidance:');
    buf.writeln('Reflections should NOT contain:');
    buf.writeln('• Patient names or identifiers');
    buf.writeln('• NHS numbers or hospital numbers');
    buf.writeln('• Dates of birth');
    buf.writeln('• Addresses or contact details');
    buf.writeln('• Specific ward/bed locations\n');
    
    buf.writeln('✅ Please use:');
    buf.writeln('• Generic descriptions (e.g., "a patient")');
    buf.writeln('• Age ranges instead of DOB');
    buf.writeln('• General locations instead of specifics');
    
    return buf.toString();
  }
}

/// Warning levels for PHI detection
enum PhiWarningLevel {
  none,     // No PHI detected
  low,      // 1 indicator - might be false positive
  medium,   // 2-3 indicators - likely contains PHI
  high,     // 4+ indicators - almost certainly contains PHI
}

extension PhiWarningLevelExt on PhiWarningLevel {
  String get description {
    switch (this) {
      case PhiWarningLevel.none:
        return 'No issues detected';
      case PhiWarningLevel.low:
        return 'Possible patient information detected';
      case PhiWarningLevel.medium:
        return 'Likely patient information detected';
      case PhiWarningLevel.high:
        return 'Patient information detected';
    }
  }
  
  bool get shouldWarn => this != PhiWarningLevel.none;
}

File: functions/src/phiDetector.ts
/**
 * PHI (Protected Health Information) Detector
 * 
 * Detects potential patient-identifiable information in text
 * for NHS GDPR compliance
 */

export interface PhiWarning {
  type: string;
  severity: 'low' | 'medium' | 'high';
  matches: string[];
}

export class PhiDetector {
  /**
   * Detect PHI in text
   */
  detectPhi(text: string): PhiWarning[] {
    const warnings: PhiWarning[] = [];

    // NHS Number pattern (10 digits with spaces)
    const nhsNumbers = this.detectNhsNumbers(text);
    if (nhsNumbers.length > 0) {
      warnings.push({
        type: 'NHS_NUMBER',
        severity: 'high',
        matches: nhsNumbers,
      });
    }

    // Names (Mr/Mrs/Ms/Dr followed by capitalized words)
    const names = this.detectNames(text);
    if (names.length > 0) {
      warnings.push({
        type: 'NAME',
        severity: 'high',
        matches: names,
      });
    }

    // Dates of birth
    const dobs = this.detectDatesOfBirth(text);
    if (dobs.length > 0) {
      warnings.push({
        type: 'DATE_OF_BIRTH',
        severity: 'high',
        matches: dobs,
      });
    }

    // Addresses
    const addresses = this.detectAddresses(text);
    if (addresses.length > 0) {
      warnings.push({
        type: 'ADDRESS',
        severity: 'medium',
        matches: addresses,
      });
    }

    // Phone numbers (UK format)
    const phoneNumbers = this.detectPhoneNumbers(text);
    if (phoneNumbers.length > 0) {
      warnings.push({
        type: 'PHONE_NUMBER',
        severity: 'medium',
        matches: phoneNumbers,
      });
    }

    // Email addresses
    const emails = this.detectEmails(text);
    if (emails.length > 0) {
      warnings.push({
        type: 'EMAIL',
        severity: 'medium',
        matches: emails,
      });
    }

    return warnings;
  }

  /**
   * Check if text contains PHI
   */
  containsPhi(text: string): boolean {
    return this.detectPhi(text).length > 0;
  }

  // Detection methods

  private detectNhsNumbers(text: string): string[] {
    // NHS numbers: 10 digits, often with spaces (e.g., 485 777 3456)
    const pattern = /\b\d{3}\s?\d{3}\s?\d{4}\b/g;
    return this.extractMatches(text, pattern);
  }

  private detectNames(text: string): string[] {
    // Titles followed by capitalized words
    const pattern = /\b(Mr|Mrs|Ms|Miss|Dr|Professor|Prof)\s+[A-Z][a-z]+(\s+[A-Z][a-z]+)*\b/g;
    return this.extractMatches(text, pattern);
  }

  private detectDatesOfBirth(text: string): string[] {
    const patterns = [
      // DD/MM/YYYY, DD-MM-YYYY
      /\b\d{1,2}[\/\-]\d{1,2}[\/\-]\d{4}\b/g,
      // DD MMM YYYY (e.g., 15 Jan 1990)
      /\b\d{1,2}\s+(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)[a-z]*\s+\d{4}\b/gi,
      // "DOB: ", "Date of Birth: "
      /\b(DOB|Date of Birth):\s*[\w\s\/\-]+/gi,
    ];

    const matches: string[] = [];
    patterns.forEach(pattern => {
      matches.push(...this.extractMatches(text, pattern));
    });
    return matches;
  }

  private detectAddresses(text: string): string[] {
    const patterns = [
      // UK postcodes
      /\b[A-Z]{1,2}\d{1,2}\s?\d[A-Z]{2}\b/gi,
      // "Address: "
      /\bAddress:\s*[^\n]+/gi,
    ];

    const matches: string[] = [];
    patterns.forEach(pattern => {
      matches.push(...this.extractMatches(text, pattern));
    });
    return matches;
  }

  private detectPhoneNumbers(text: string): string[] {
    const patterns = [
      // UK phone numbers
      /\b0\d{10}\b/g, // 07123456789
      /\b0\d{3}\s?\d{3}\s?\d{4}\b/g, // 0123 456 7890
      /\b\+44\s?\d{10}\b/g, // +44 7123456789
    ];

    const matches: string[] = [];
    patterns.forEach(pattern => {
      matches.push(...this.extractMatches(text, pattern));
    });
    return matches;
  }

  private detectEmails(text: string): string[] {
    const pattern = /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/g;
    return this.extractMatches(text, pattern);
  }

  /**
   * Extract unique matches from text using regex
   */
  private extractMatches(text: string, pattern: RegExp): string[] {
    const matches = text.match(pattern);
    if (!matches) return [];
    
    // Return unique matches
    return [...new Set(matches)];
  }
}

export const phiDetector = new PhiDetector();


